### PR TITLE
Teacher forcing

### PR DESCRIPTION
**Pull request by Katsutoshii**
Removes autoregressive loop from model.call into model.call_n.
This reduces model memory requirements by a factor of roughly storm_duration / N_FLOOD_MAPS because we no longer predict the entire sequence of flood maps, which unlocks GPU training.

### CODE REVIEW

**Comment by waltaskew** [comment]
Let's put together a smaller PR.

This contains more code than we want to review at a time. It also includes some code we know we don't want to merge as is, such as `metastore.py` and `featurelabelchunks.py` (those two files I rewrote as https://github.com/UrbanSystemsLab/climateiq-cnn/pull/112)

Can you put together a PR with just the changes to the model itself to start? The dataset generation we'll want to work on before reviewing & merging.

**Comment by Katsutoshii** [comment]
> Let's put together a smaller PR.
> 
> This contains more code than we want to review at a time. It also includes some code we know we don't want to merge as is, such as `metastore.py` and `featurelabelchunks.py` (those two files I rewrote as #112)
> 
> Can you put together a PR with just the changes to the model itself to start? The dataset generation we'll want to work on before reviewing & merging.

Sure, I've reverted the extra files. The branch with e2e training colab is now at https://github.com/Katsutoshii/climateiq-cnn/tree/teacher-forcing-e2e

File: usl_models/usl_models/flood_ml/model.py
Comment (waltaskew) [inline usl_models/usl_models/flood_ml/model.py:133]: Looks like this is accidentally commented out?

File: usl_models/usl_models/flood_ml/model.py
Comment (tatianawu) [inline usl_models/usl_models/flood_ml/model.py:302]: Can we maybe add a comment saying this is the channel dimension for the spatiotemporal tensors? Unlike the other variables, there's no context to infer the variable meaning.

File: usl_models/usl_models/flood_ml/model.py
Comment (tatianawu) [inline usl_models/usl_models/flood_ml/model.py:343]: Need to update the name here to `full_input`. Can you also explain how this differs from the `input` in the `call` method?

File: usl_models/usl_models/flood_ml/model_params.py
Comment (tatianawu) [inline usl_models/usl_models/flood_ml/model_params.py:27]: Will this create a default Optimizer that's passed by reference to all instances if it's not overridden?

I switched to the string value in this [commit](https://github.com/UrbanSystemsLab/climateiq-cnn/pull/93/commits/62c5036ef0a2018337185f123204991e5f4bce25) to avoid possible pass-by-reference issues where the same optimizer gets shared across multiple models, which can cause downstream failures.

File: usl_models/usl_models/flood_ml/model.py
Comment (tatianawu) [inline usl_models/usl_models/flood_ml/model.py:386]: Is this a replacement for the `temporal_window_view` function in `data_utils.py`? If so, can you add more detail to the docstring explaining what this is doing? It would also be useful to add a test to make sure the behavior is WAI (and will help others understand what inputs/outputs might look like.

**Review by tatianawu — COMMENTED** [review]
Did you update the tests based on these changes? Pytest didn't run with the most recent update, but I'd expect some of the tests to no longer be relevant.

File: usl_models/usl_models/flood_ml/model.py
Comment (waltaskew) [inline usl_models/usl_models/flood_ml/model.py:124]: what does logging into a 'logs' directory mean in the context of VertexAI training? The logs directory would be in an ephemeral docker container.
This looks like more of a debug option than a default option?

File: usl_models/usl_models/flood_ml/model.py
Comment (Katsutoshii) [reply usl_models/usl_models/flood_ml/model.py]: Correct, I've deprecated that from _model_params to be consistent with keras API.

File: usl_models/usl_models/flood_ml/model.py
Comment (Katsutoshii) [reply usl_models/usl_models/flood_ml/model.py]: Sure, I've moved this callback outside of model.fit.

File: usl_models/usl_models/flood_ml/model_params.py
Comment (Katsutoshii) [reply usl_models/usl_models/flood_ml/model_params.py]: It seems limiting to not be able to specify the parameters of Adam (learning rate etc). Using the default factor avoids these issues as long as the client does not share the Params across multiple models.

Comment (tatianawu) [reply usl_models/usl_models/flood_ml/model_params.py]: The default value doesn't limit anything? It's just there if the user doesn't provide their own optimizer, in which they can specify whatever parameters they want. I have the string as a default to prevent any accidental pass-by-reference issues when the user _doesn't_ specify an optimizer. But your optimizer is equivalent to mine.

File: usl_models/usl_models/flood_ml/model.py
Comment (Katsutoshii) [reply usl_models/usl_models/flood_ml/model.py]: Fixed.

File: usl_models/usl_models/flood_ml/model.py
Comment (Katsutoshii) [reply usl_models/usl_models/flood_ml/model.py]: I've moved this inside the class, now it's FloodModel.Input.

File: usl_models/usl_models/flood_ml/model.py
Comment (Katsutoshii) [reply usl_models/usl_models/flood_ml/model.py]: Updated.

File: usl_models/usl_models/flood_ml/model.py
Comment (Katsutoshii) [reply usl_models/usl_models/flood_ml/model.py]: Suggestion applied.

File: usl_models/usl_models/flood_ml/model.py
Comment (Katsutoshii) [reply usl_models/usl_models/flood_ml/model.py]: Done.

File: usl_models/usl_models/flood_ml/model.py
Comment (Katsutoshii) [reply usl_models/usl_models/flood_ml/model.py]: Done.

File: usl_models/usl_models/flood_ml/model.py
Comment (Katsutoshii) [reply usl_models/usl_models/flood_ml/model.py]: Yes, I've added a unit test for this.

File: usl_models/tests/__init__.py
Comment (waltaskew) [inline usl_models/tests/__init__.py:1]: We shoudn't need all the empty `__init__.py`s in python3.
I think I'd actually prefer not to have one here in `usl_models/tests/__init__.py` because it marks it as part of the package installed by `setup.py`.

File: usl_models/usl_models/flood_ml/dataset.py
Comment (waltaskew) [inline usl_models/usl_models/flood_ml/dataset.py:175]: What does this branch protect against? The batch_size being 0?

File: usl_models/tests/__init__.py
Comment (Katsutoshii) [reply usl_models/tests/__init__.py:1]: I'm on latest pytest (8.2.2) and Python 3.10.14 and the tests are unable to import any of our libraries unless I include the __init__.py files.

File: usl_models/usl_models/flood_ml/dataset.py
Comment (waltaskew) [inline usl_models/usl_models/flood_ml/dataset.py:19]: It looks like docstrings & comments are copy-pasted from `load_dataset_windowed`. They reference window-ing, which looks specific to `load_dataset_windowed`.

It would be helpful to explain in the docstrings when a caller should use `load_dataset` vs. `load_dataset_windowed`. Why do we need both?

We should also have a test for this function.

It may be easier to add this in another PR since it doesn't look like this function is used at present.

File: usl_models/tests/__init__.py
Comment (waltaskew) [reply usl_models/tests/__init__.py:1]: init files in the test modules shouldn't impact the tests' abilities to import other files
it isn't a big deal to include them, but I'm a little curious to understand the issue you're having

File: usl_models/usl_models/flood_ml/dataset.py
Comment (Katsutoshii) [reply usl_models/usl_models/flood_ml/dataset.py:175]: Yes, when batch size is zero it avoids adding an extra batch dimension, which was what caused the batch prediction error.

Comment (waltaskew) [reply usl_models/usl_models/flood_ml/dataset.py:175]: Okay! A comment to that effect would be helpful.

Comment (waltaskew) [reply usl_models/usl_models/flood_ml/dataset.py:174]: a test for that case would also be helpful so we don't hit this error again!

File: usl_models/usl_models/flood_ml/dataset.py
Comment (Katsutoshii) [reply usl_models/usl_models/flood_ml/dataset.py:19]: Fixed.

File: usl_models/usl_models/flood_ml/model_params.py
Comment (tatianawu) [inline usl_models/usl_models/flood_ml/model_params.py:40]: Walt had initially asked for this to be removed from this file and into the `tests/` directory. Is there a reason you moved it back here?

File: usl_models/tests/flood_ml/flood_model_test.py
Comment (tatianawu) [inline usl_models/tests/flood_ml/flood_model_test.py:51]: I think this comment and the docstring is slightly outdated. Can you update?

File: usl_models/tests/flood_ml/flood_model_test.py
Comment (tatianawu) [inline usl_models/tests/flood_ml/flood_model_test.py:160]: This basically seems to just replicate the code from the actual method. Maybe it would be useful to create observable tensors? This has the benefit of also helping users understand what exactly the method is doing.

I'm picturing something similar to https://github.com/UrbanSystemsLab/climateiq-cnn/blob/main/usl_models/tests/atmo_ml/atmo_utils_test.py.

File: usl_models/usl_models/flood_ml/model.py
Comment (tatianawu) [inline usl_models/usl_models/flood_ml/model.py:9]: Nit: Can we alphabetize?

File: usl_models/usl_models/flood_ml/model.py
Comment (tatianawu) [reply usl_models/usl_models/flood_ml/model.py]: Can you add details to the docstring explaining what the `temporal` tensor looks like when this is called?

Let's add the args and return section; I think it's better to have more detail than less here. The "n-sized window" is a bit confusing, because there are actually two temporal windows of interest in the model: the rainfall `m` and the spatiotemporal `n`. We're actually retrieving the `n` most recent (`m`-sized) rainfall windows at the current time step.

FWIW I wrote the original code and this still trips me up.

File: usl_models/tests/__init__.py
Comment (tatianawu) [reply usl_models/tests/__init__.py:1]: Did you run `pip install` in your local environment? You'll want:

1. `pip install -r requirements.txt` to make sure you're using the same versions
2. `pip install -e .[dev]` to install the local project

File: usl_models/tests/flood_ml/flood_model_test.py
Comment (Katsutoshii) [reply usl_models/tests/flood_ml/flood_model_test.py:62]: Done

File: usl_models/usl_models/flood_ml/model.py
Comment (Katsutoshii) [reply usl_models/usl_models/flood_ml/model.py]: Done.

File: usl_models/tests/flood_ml/flood_model_test.py
Comment (Katsutoshii) [reply usl_models/tests/flood_ml/flood_model_test.py]: Done.

File: usl_models/usl_models/flood_ml/model.py
Comment (Katsutoshii) [reply usl_models/usl_models/flood_ml/model.py:9]: Done.

File: usl_models/usl_models/flood_ml/model_params.py
Comment (Katsutoshii) [reply usl_models/usl_models/flood_ml/model_params.py]: This was a merge snafu. Reverted.

File: usl_models/usl_models/flood_ml/dataset.py
Comment (Katsutoshii) [reply usl_models/usl_models/flood_ml/dataset.py:174]: Separated this into another PR.

File: usl_models/tests/flood_ml/flood_model_test.py
Comment (tatianawu) [inline usl_models/tests/flood_ml/flood_model_test.py:178]: Checking my understanding here. With `t = 3`, we have 2 rows of zero padding, and the data retrieved from `temporal` are rows 0-2, correct?

Could we set temporal to 1-8 instead? That row of zeros might be confusing.

**Review by tatianawu — APPROVED** [review]
LGTM. I ignored the dataset files. Will all of these be rebased away, or are there any I should take a look at?

**Review by SallyElHajjar — APPROVED** [review]

**Review by waltaskew — APPROVED** [review]

**Pull request merged**
